framework/src/timeintegrators/CrankNicolson.C

```diff
@@ -42,21 +42,17 @@ CrankNicolson::computeTimeDerivatives()
 }
 
 void
-CrankNicolson::preSolve()
+CrankNicolson::init()
 {
-  if (_t_step == 1)
-  {
-    // make sure that time derivative contribution is zero in the first pre-solve step
-    _u_dot.zero();
-    _u_dot.close();
+  // make sure that time derivative contribution is zero in the first pre-solve step
+  _u_dot.zero();
+  _u_dot.close();
 
-    _du_dot_du = 0;
+  _du_dot_du = 0;
 
-    // for the first time step, compute residual for the old time step
-    _fe_problem.computeResidualType(_solution_old, _nl.RHS(), Moose::KT_NONTIME);
-    _residual_old = _nl.RHS();
-    _residual_old.close();
-  }
+  // compute residual for the initial time step
+  _fe_problem.computeResidualType(*_solution, _nl.RHS(), Moose::KT_NONTIME);
```

Comment:
I will look into this.  I forgot why `_nl.RHS()` was needed...

Reply:
Here is why you cannot do this.  We are passing `_residual_old` as a residual when calling `computeResidualType`. Then in `NonlinearSystemBase::computeResidual` around line 550, we have:
```
    if (_time_integrator)
      _time_integrator->postResidual(residual);
    else
      residual += *_Re_non_time;
```
So we call the `postResidual` method and the will try to do (`CrankNicolson.C:L62`):
```
  residual += _residual_old;
```
But `residual` is `_residual_old` and that why we get the PETSc error.

Reply:
This patch:
```
--- a/framework/src/base/NonlinearSystemBase.C
+++ b/framework/src/base/NonlinearSystemBase.C
@@ -548,8 +548,10 @@ NonlinearSystemBase::computeResidual(NumericVector<Number> & residual, Moose::Ke
     if (_Re_time)
       _Re_time->close();
     _Re_non_time->close();
-    if (_time_integrator)
+    if (_time_integrator && type == Moose::KT_ALL)
       _time_integrator->postResidual(residual);
+    else if (_Re_time && type == Moose::KT_TIME)
+      residual += *_Re_time;
     else
       residual += *_Re_non_time;
     residual.close();
```
Should allow you to use `_residual_old` instead `_nl.RHS()`.

Reply:
Bypassing `postResidual` looks dangerous. With the reason in mind, maybe we can just check `if (&residual == &_residual_old)` in `CrankNicolson::postResidual()`? Or just keep the way it is with a comment why we are doing it.

Reply:
Let's think about it. `postResidual` is valid only if we form `KT_ALL`, because it means to form TIME and NONTIME part and combine them, which is what `postResidual` is responsible for.

If I am forming just KT_TIME, then my final residual includes just the TIME part, no need to combine it with non-time, because I do not care about the non-time part. The same holds for NON_TIME.

Reply:
There could be logic issue here. You check `_Re_time` in else if, but `CrankNicolson::postResidual` add it regardless. So are we assuming `_Re_time` will be valid if there is a time integrator?

Reply:
If we move `residual += _residual_old;` above the other two lines, will the PETSc error be gone? I am in middle of something, will take some time for me to figure it out on my own. Maybe your guys have a quicker answer? I am asking because I think the logic here is too complicated and probably not final. I want to make this PR less intrusive.

Reply:
> So are we assuming _Re_time will be valid if there is a time integrator?

Reply:
Maybe at the beginning, vector is not in assembly stage, thus it can add on top itself? Just curious.

Reply:
Nope. It is just not possible in PETSc, see [the code](https://github.com/petsc/petsc/blob/96f3724d6c194183abb724041525c35f89c72dc7/src/vec/vec/interface/rvector.c#L606)

Reply:
I see. How about I leave this as it is? Vector assignment should not cost much. I just feel the logic is too complicated. We possibly can do better on this. It could be also related with the kernel tagging stuff.

Reply:
Works for me.